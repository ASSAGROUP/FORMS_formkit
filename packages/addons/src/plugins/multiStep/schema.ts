import { FormKitSchemaNode } from '@formkit/core'

export const multiStepSchema: FormKitSchemaNode[] = [
  {
    $el: 'div',
    attrs: {
      'data-type': 'multi-step',
      class: '$classes.outer',
    },
    children: [
      {
        $el: 'ul',
        attrs: {
          class: '$classes.tabs',
        },
        children: [
          {
            $el: 'li',
            for: ['step', 'key', '$steps'],
            attrs: {
              class: '$classes.tab',
              onClick: '$handlers.setActiveStep($step)',
            },
            children: ['$step.name'],
          },
        ],
      },
      {
        $el: 'div',
        attrs: {
          class: '$classes.wrapper',
        },
        children: [
          {
            $el: 'div',
            attrs: {
              class: '$classes.inner',
            },
            children: '$slots.default',
          },
        ],
      },
    ],
  },
]

export const stepSchema: FormKitSchemaNode[] = [
  {
    $el: 'div',
    attrs: {
      'data-type': 'step',
      class: '$classes.step',
      style: {
        if: '$isActiveStep',
        then: '',
        else: 'display: none;',
      },
    },
    children: [
      {
        $el: 'div',
        attrs: {
          class: '$classes.stepInner',
        },
        children: '$slots.default',
      },
      {
        $el: 'div',
        attrs: {
          class: '$classes.stepActions',
        },
        children: [
          {
            $el: 'div',
            if: '$: !$isFirstStep',
            attrs: {
              class: '$classes.actionPrevious',
            },
            children: [
              {
                $formkit: 'button',
                label: {
                  if: '$prevLabel',
                  then: '$prevLabel',
                  else: 'Previous',
                },
              },
            ],
          },
          {
            $el: 'div',
            attrs: {
              class: '$classes.actionNext',
            },
            children: [
              {
                $formkit: 'button',
                label: {
                  if: '$nextLabel',
                  then: '$nextLabel',
                  else: 'Next',
                },
              },
            ],
          },
        ],
      },
    ],
  },
]
