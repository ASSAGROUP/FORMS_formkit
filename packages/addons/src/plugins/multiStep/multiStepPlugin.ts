import {
  FormKitNode,
  FormKitPlugin,
  createMessage,
  FormKitFrameworkContext,
} from '@formkit/core'
import { multiStep, step } from './schema'

/**
 * The options to be passed to {@link createMultiStepPlugin | createMultiStepPlugin}
 *
 * @public
 */
export interface MultiStepOptions {
  // flattenValues?: boolean
  allowIncomplete?: boolean
  hideProgressLabels?: boolean
  tabStyle?: 'tab' | 'progress'
}

type FormKitFrameworkContextWithSteps = FormKitFrameworkContext & {
  steps: FormKitFrameworkContext[]
  stepIndex: number
}

/**
 * Coverts a camelCase string to a title case string
 *
 * @param str - The string to convert
 * @returns string
 */
const camel2title = (str: string) =>
  str
    .replace(/([A-Z])/g, (match: string) => ` ${match}`)
    .replace(/^./, (match: string) => match.toUpperCase())
    .trim()

/**
 * Iterates through each step and sets props to help
 * determine step poisitioning within the multi-step.
 *
 * @param steps - The steps to iterate through
 */
function setNodePositionProps(steps: FormKitFrameworkContext[]) {
  steps.forEach((step: FormKitFrameworkContext, index: number) => {
    step.isFirstStep = index === 0
    step.isLastStep = index === steps.length - 1
    step.stepIndex = index
    step.steps = steps
  })
}

function showStepErrors(step: FormKitFrameworkContext) {
  if (!step.showStepErrors) return
  return (
    parseInt(step.blockingCount as string) +
    parseInt(step.errorCount as string) >
    0
  )
}

/**
 * Determines if the target step can be navigated to based on current
 * configuration options and the state of the current step.
 *
 * @param currentStep - The current step
 * @param targetStep - The target step
 */
function isTargetStepAllowed(
  currentStep: FormKitFrameworkContext,
  targetStep: FormKitFrameworkContext
) {
  const { allowIncomplete } = currentStep.node.parent?.props || {}
  const parentNode = currentStep.node.parent
  const currentStepIndex = parentNode?.props.steps.indexOf(currentStep)
  const targetStepIndex = parentNode?.props.steps.indexOf(targetStep)

  // check if there is a function for the stepChange guard
  const beforeStepChange = currentStep.node.props.beforeStepChange
    || currentStep.node.parent?.props.beforeStepChange

  if (beforeStepChange && typeof beforeStepChange === 'function') {
    const result = beforeStepChange({
      currentStep,
      targetStep,
      direction: targetStepIndex < currentStepIndex ? 'backwards' : 'forwards'
    });
    if (typeof result === 'boolean' && !result) return false;
  }

  if (targetStepIndex < currentStepIndex) {
    // we can always step backwards
    return true
  }

  // check how many steps we need to step forward
  // and then check that each intermediate step is valid
  const delta = targetStepIndex - currentStepIndex
  for (let i = 0; i < delta; i++) {
    const intermediateStep = parentNode?.props.steps[currentStepIndex + i]
    if (i === 0) {
      // trigger any applicable validation errors on the first intermediate step
      triggerStepValidations(intermediateStep)
      intermediateStep.showStepErrors = true
    }
    const stepIsAllowed = allowIncomplete || intermediateStep.state?.valid
    if (!stepIsAllowed) {
      return false
    }
  }

  return true
}

/**
 * Changes the active step to the target step if the target step is allowed.
 *
 * @param targetStep - The target step
 */
function setActiveStep(targetStep: FormKitFrameworkContext, e: Event) {
  e.preventDefault()
  if (targetStep && targetStep.node.name && targetStep.node.parent) {
    const currentStep = targetStep.node.parent.props.steps.find(
      (step: FormKitFrameworkContext) =>
        step.node.name === targetStep.node.parent?.props.activeStep
    )
    const stepIsAllowed = isTargetStepAllowed(currentStep, targetStep)

    if (stepIsAllowed && targetStep.node.parent.context) {
      targetStep.node.parent.props.activeStep = targetStep.node.name
    }
  }
}

/**
 * Changes the current step by the delta value if the target step is allowed.
 *
 * @param delta - The number of steps to increment or decrement
 * @param step - The current step
 */
function incrementStep(
  delta: number,
  currentStep: FormKitFrameworkContextWithSteps
) {
  if (currentStep && currentStep.node.name && currentStep.node.parent) {
    const {
      steps,
      stepIndex,
    }: { steps: FormKitFrameworkContext[]; stepIndex: number } = currentStep
    const targetStep = steps[stepIndex + delta]
    const stepIsAllowed = isTargetStepAllowed(currentStep, targetStep)

    if (targetStep && stepIsAllowed) {
      currentStep.node.parent.props.activeStep = targetStep.node.name
    }
  }
}

/**
 * Causes the display of any validation errors on the target step.
 *
 * @param step - The current step
 * @returns Boolean
 */
function triggerStepValidations(step: FormKitFrameworkContext) {
  step.node.walk((n) => {
    n.store.set(
      createMessage({
        key: 'submitted',
        value: true,
        visible: false,
      })
    )
  })
  return (
    step.node.context?.state.valid || step.node.parent?.props.allowIncomplete
  )
}

/**
 * Creates a new multi-step plugin.
 *
 * @param options - The options of {@link MultiStepOptions | MultiStepOptions} to pass to the plugin
 *
 * @returns A {@link @formkit/core#FormKitPlugin | FormKitPlugin}
 *
 * @public
 */
export function createMultiStepPlugin(
  options?: MultiStepOptions
): FormKitPlugin {
  const multiStepPlugin = (node: FormKitNode) => {
    if (node.props.type === 'multi-step') {
      node.addProps([
        'steps',
        'activeStep',
        'beforeStepChange',
        'flattenValues',
        'allowIncomplete',
        'hideProgressLabels',
        'tabStyle',
      ])

      // determine correct default prop values
      // node.props.flattenValues =
      //   typeof node.props.flattenValues === 'boolean'
      //     ? node.props.flattenValues
      //     : options?.flattenValues || false
      node.props.allowIncomplete =
        typeof node.props.allowIncomplete === 'boolean'
          ? node.props.allowIncomplete
          : options?.allowIncomplete || false
      node.props.hideProgressLabels =
        typeof node.props.hideProgressLabels === 'boolean'
          ? node.props.hideProgressLabels
          : options?.hideProgressLabels || false
      node.props.tabStyle = node.props.tabStyle || options?.tabStyle || 'tab'

      node.on('created', () => {
        if (!node.context) return
        node.context.handlers.triggerStepValidations = triggerStepValidations
        node.context.handlers.showStepErrors = showStepErrors
      })

      node.on('childRemoved', ({ payload: childNode }) => {
        let removedStepIndex = -1
        node.props.steps = node.props.steps.filter(
          (step: FormKitFrameworkContext, index: number) => {
            if (step.node.name !== childNode.name) {
              return true
            }
            removedStepIndex = index
            return false
          }
        )
        // if the child that was removed was the active step
        // then fallback to the next available step
        if (node.props.activeStep === childNode.name) {
          const targetIndex = removedStepIndex > 0 ? removedStepIndex - 1 : 0
          node.props.activeStep = node.props.steps[targetIndex]
            ? node.props.steps[targetIndex].node.name
            : ''
        }

        // recompute step positions
        setNodePositionProps(node.props.steps)
      })
    }
    if (
      node.props.type === 'step' &&
      node.parent?.props.type === 'multi-step'
    ) {
      node.addProps([
        'isActiveStep',
        'isFirstStep',
        'isLastStep',
        'beforeStepChange',
        'stepName',
        'errorCount',
        'blockingCount',
        'totalErrorCount',
        'showStepErrors',
        'isValid',
        'hasBeenVisited',
      ])
      node.on('created', () => {
        if (!node.context) return
        if (node.parent && node.parent.context) {
          node.props.stepName = node.props.label || camel2title(node.name)
          node.props.errorCount = 0
          node.props.blockingCount = 0

          const parentNode = node.parent
          parentNode.props.steps = Array.isArray(parentNode.props.steps)
            ? [...parentNode.props.steps, node.context]
            : [node.context]

          setNodePositionProps(parentNode.props.steps)

          parentNode.props.activeStep = parentNode.props.activeStep
            ? parentNode.props.activeStep
            : parentNode.props.steps[0]
              ? parentNode.props.steps[0].node.name
              : ''

          if (parentNode.context) {
            parentNode.context.handlers.setActiveStep = (
              stepNode: FormKitFrameworkContext
            ) => setActiveStep.bind(null, stepNode)
            node.context.handlers.incrementStep = (
              delta: number,
              stepNode: FormKitFrameworkContextWithSteps
            ) => incrementStep.bind(null, delta, stepNode)
          }
        }
      })

      node.on('count:errors', ({ payload: count }) => {
        node.props.errorCount = count
      })
      node.on('count:blocking', ({ payload: count }) => {
        node.props.blockingCount = count
      })

      function updateTotalErrorCount(node: FormKitNode) {
        node.props.totalErrorCount =
          node.props.errorCount + node.props.blockingCount
      }
      node.on('prop:errorCount', () => updateTotalErrorCount(node))
      node.on('prop:blockingCount', () => updateTotalErrorCount(node))
      node.on('prop:totalErrorCount', () => {
        node.props.isValid = node.props.totalErrorCount <= 0
      })

      node.on('message-added', ({ payload }) => {
        if (payload.key === 'submitted') {
          updateTotalErrorCount(node)
          if (node.context) {
            triggerStepValidations(node.context)
            node.props.showStepErrors = true
          }
        }
      })

      node.parent.on('prop:activeStep', ({ payload }) => {
        node.props.isActiveStep = node.name === payload
      })
      node.on('prop:isActiveStep', () => {
        if (!node.props.hasBeenVisited && node.props.isActiveStep) {
          node.props.hasBeenVisited = true
        }
      })
    } else if (node.parent?.props.type === 'multi-step') {
      console.warn(
        'Invalid FormKit input location. <FormKit type="multi-step"> should only have <FormKit type="step"> inputs as immediate children. Failure to wrap child inputs in <FormKit type="step"> can lead to undesired behaviors.'
      )
    }
  }

  multiStepPlugin.library = (node: FormKitNode) => {
    switch (node.props.type) {
      case 'multi-step':
        return node.define(multiStep)
      case 'step':
        const isInvalid =
          !node.parent || node.parent.props.type !== 'multi-step'
        if (isInvalid) {
          console.warn(
            'Invalid use of <FormKit type="step">. <FormKit type="step"> must be an immediate child of a <FormKit type="multi-step"> component.'
          )
        }
        return node.define(step)
    }
  }

  return multiStepPlugin
}
