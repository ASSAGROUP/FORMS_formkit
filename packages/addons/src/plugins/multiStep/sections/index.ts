export { stepActions } from './stepActions'
export { stepNext } from './stepNext'
export { stepOuter } from './stepOuter'
export { stepPrevious } from './stepPrevious'
export { tab } from './tab'
export { tabs } from './tabs'
