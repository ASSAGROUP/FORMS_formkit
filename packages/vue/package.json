{
  "name": "@formkit/vue",
  "repository": "https://github.com/formkit/formkit",
  "version": "1.0.0-beta.1",
  "description": "Build industry leading Vue forms 10x faster.",
  "main": "dist/index.cjs.js",
  "module": "dist/index.esm.js",
  "types": "dist/index.d.ts",
  "author": "Justin Schroeder <justin@wearebraid.com>",
  "license": "UNLICENSED",
  "publishConfig": {
    "access": "restricted",
    "registry": "https://npm.pkg.github.com"
  },
  "scripts": {
    "test": "jest"
  },
  "type": "module",
  "dependencies": {
    "@formkit/core": "1.0.0-beta.1",
    "@formkit/i18n": "1.0.0-beta.1",
    "@formkit/inputs": "1.0.0-beta.1",
    "@formkit/observer": "1.0.0-beta.1",
    "@formkit/rules": "1.0.0-beta.1",
    "@formkit/utils": "1.0.0-beta.1",
    "@formkit/validation": "1.0.0-beta.1"
  },
  "devDependencies": {
    "@vue/test-utils": "^2.0.0-rc.14"
  },
  "peerDependencies": {
    "vue": "^3.2.1"
  }
}