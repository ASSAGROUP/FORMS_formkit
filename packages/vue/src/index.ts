/**
 * The useInput composable.
 * @public
 */
export { useInput } from './composables/useInput'

/**
 * Shorthand for creating inputs with standard FormKit features.
 */
export { createInput } from './composables/createInput'

/**
 * A special watcher for Vue that reports the location of a deep mutation.
 */
export { default as watchVerbose } from './composables/watchVerbose'

/**
 * A special watcher for Vue that reports the location of a deep mutation.
 */
export { isObserver } from './composables/mutationObserver'

/**
 * The plugin and plugin types.
 * @public
 */
export * from './plugin'

/**
 * The root FormKit component.
 * @public
 */
export { default as FormKit, parentSymbol } from './FormKit'

/**
 * The FormKitSchema component.
 * @public
 */
export { FormKitSchema } from './FormKitSchema'

/**
 * The default configuration.
 * @public
 */
export { default as defaultConfig, DefaultConfigOptions } from './defaultConfig'

/**
 * The vue specific FormKit core plugin. This is generally required for all
 * vue based FormKit configurations.
 * @public
 */
export { default as bindings } from './bindings'

/**
 * Export the reset count explicitly
 */
export {
  resetCount,
  errorHandler,
  setErrors,
  submitForm,
  reset,
} from '@formkit/core'
