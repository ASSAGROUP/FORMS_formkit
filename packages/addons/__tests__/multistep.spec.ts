import { h } from 'vue'
import { mount } from '@vue/test-utils'
import { FormKit, FormKitSchema, plugin, defaultConfig } from '@formkit/vue'
import { createMultiStepPlugin } from '../src/plugins/multiStep/multiStepPlugin'
import { jest } from '@jest/globals'

const multiStepSchemaBasic = [
  {
    $formkit: 'multi-step',
    children: [
      {
        $formkit: 'step',
        name: 'stepOne',
        children: [
          {
            $formkit: 'text',
            validation: 'required',
          },
        ],
      },
      {
        $formkit: 'step',
        name: 'stepTwo',
      },
      {
        $formkit: 'step',
        name: 'stepThree',
      },
    ],
  },
]

const multiStepSchemaBasicWithProps = [
  {
    $formkit: 'multi-step',
    tabStyle: 'progress',
    hideProgressLabels: true,
    allowIncomplete: true,
    children: [
      {
        $formkit: 'step',
        name: 'stepOne',
        children: [
          {
            $formkit: 'text',
            validation: 'required',
          },
        ],
      },
      {
        $formkit: 'step',
        name: 'stepTwo',
      },
      {
        $formkit: 'step',
        name: 'stepThree',
      },
    ],
  },
]

describe('multistep', () => {
  afterEach(() => {
    // restore the spy created with spyOn
    jest.restoreAllMocks()
  })

  it('it can mount a multi-step input', () => {
    const wrapper = mount(FormKit, {
      props: {
        type: 'multi-step',
      },
      global: {
        plugins: [
          [
            plugin,
            defaultConfig({
              plugins: [createMultiStepPlugin()],
            }),
          ],
        ],
      },
    })

    expect(wrapper.html()).toContain('data-type="multi-step"')
  })

  it('it can mount steps inside of a multistep input', () => {
    const wrapper = mount(FormKit, {
      props: {
        type: 'multi-step',
      },
      global: {
        plugins: [
          [
            plugin,
            defaultConfig({
              plugins: [createMultiStepPlugin()],
            }),
          ],
        ],
      },
      slots: {
        default() {
          return h(FormKit, {
            type: 'step',
          })
        },
      },
    })
    expect(wrapper.html()).toContain('data-type="step"')
  })

  it('it throws a warning when using a step input outside of a multi-step', () => {
    jest.spyOn(global.console, 'warn').mockImplementation(() => {})

    mount(FormKit, {
      props: {
        type: 'step',
      },
      global: {
        plugins: [
          [
            plugin,
            defaultConfig({
              plugins: [createMultiStepPlugin()],
            }),
          ],
        ],
      },
    })
    expect(console.warn).toHaveBeenCalledTimes(1)
    expect(console.warn).toBeCalledWith(
      'Invalid use of <FormKit type="step">. <FormKit type="step"> must be an immediate child of a <FormKit type="multi-step"> component.'
    )
  })

  it('it throws a warning when using a non-step input as a direct child of a multi-step', () => {
    jest.spyOn(global.console, 'warn').mockImplementation(() => {})

    mount(FormKit, {
      props: {
        type: 'multi-step',
      },
      global: {
        plugins: [
          [
            plugin,
            defaultConfig({
              plugins: [createMultiStepPlugin()],
            }),
          ],
        ],
      },
      slots: {
        default() {
          return h(FormKit, {
            type: 'text',
          })
        },
      },
    })
    expect(console.warn).toHaveBeenCalledTimes(1)
    expect(console.warn).toBeCalledWith(
      'Invalid FormKit input location. <FormKit type="multi-step"> should only have <FormKit type="step"> inputs as immediate children. Failure to wrap child inputs in <FormKit type="step"> can lead to undesired behaviors.'
    )
  })

  it('creates 3 step tabs when it has 3 children of type step', async () => {
    const wrapper = mount(FormKitSchema, {
      props: {
        schema: multiStepSchemaBasic,
      },
      global: {
        plugins: [
          [
            plugin,
            defaultConfig({
              plugins: [createMultiStepPlugin()],
            }),
          ],
        ],
      },
    })

    await new Promise((r) => setTimeout(r, 5))
    expect(wrapper.html().match(/button class=\"formkit-tab\"/g)?.length).toBe(
      3
    )
  })

  it('defaults to expected prop arguments', async () => {
    const wrapper = mount(FormKitSchema, {
      props: {
        schema: multiStepSchemaBasic,
      },
      global: {
        plugins: [
          [
            plugin,
            defaultConfig({
              plugins: [createMultiStepPlugin()],
            }),
          ],
        ],
      },
    })

    await new Promise((r) => setTimeout(r, 5))
    expect(wrapper.html()).toContain('data-tab-style="tab"')
    expect(wrapper.html()).toContain('data-hide-labels="false"')
  })

  it('accepts new prop argument values', async () => {
    const wrapper = mount(FormKitSchema, {
      props: {
        schema: multiStepSchemaBasicWithProps,
      },
      global: {
        plugins: [
          [
            plugin,
            defaultConfig({
              plugins: [createMultiStepPlugin()],
            }),
          ],
        ],
      },
    })

    await new Promise((r) => setTimeout(r, 5))
    expect(wrapper.html()).toContain('data-tab-style="progress"')
    expect(wrapper.html()).toContain('data-hide-labels="true"')
  })

  it('Does not allow step advancment when current step is invalid', async () => {
    const wrapper = mount(FormKitSchema, {
      props: {
        schema: multiStepSchemaBasic,
      },
      global: {
        plugins: [
          [
            plugin,
            defaultConfig({
              plugins: [
                createMultiStepPlugin({
                  allowIncomplete: false,
                }),
              ],
            }),
          ],
        ],
      },
    })

    await new Promise((r) => setTimeout(r, 5))
    wrapper.find('.formkit-step-next button').trigger('click')
    await new Promise((r) => setTimeout(r, 15))
    expect(wrapper.html()).toContain(
      '<div class="formkit-tabs" role="tablist"><button class="formkit-tab" type="button" data-active="true" data-valid="false" data-visited="true" role="tab" id="input_33_tab_0" aria-selected="true" aria-controls="input_34" tabindex="0"><span class="formkit-tab-label">Step One</span><span class="formkit-badge" role="presentation">1</span>'
    )
  })

  it('Allows step advancment when current step is invalid but allowIncomplete is true', async () => {
    const wrapper = mount(FormKitSchema, {
      props: {
        schema: multiStepSchemaBasicWithProps,
      },
      global: {
        plugins: [
          [
            plugin,
            defaultConfig({
              plugins: [
                createMultiStepPlugin({
                  allowIncomplete: true,
                }),
              ],
            }),
          ],
        ],
      },
    })

    await new Promise((r) => setTimeout(r, 5))
    wrapper.find('.formkit-step-next button').trigger('click')
    await new Promise((r) => setTimeout(r, 15))
    // 2nd tab is active (without labels due to props)
    expect(wrapper.html()).toContain(
      '</button><button class="formkit-tab" type="button" data-active="true" data-valid="true" role="tab" id="input_42_tab_1" aria-selected="true" aria-controls="input_45" tabindex="0" data-visited="true">'
    )
    // 2nd step is visible
    expect(wrapper.html()).toContain(
      `</div>
      <div class=\"formkit-step\" data-type=\"step\" id=\"input_45\" role=\"tabpanel\" aria-labelledby=\"input_42_tab_1\">`
    )
  })
})
