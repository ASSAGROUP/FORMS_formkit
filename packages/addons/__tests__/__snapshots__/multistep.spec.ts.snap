// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`multistep Allows step advancment when current step is invalid but allowIncomplete is true 1`] = `
"<div class=\\"formkit-outer\\" id=\\"input_30\\" data-family=\\"multi-step\\" data-type=\\"multi-step\\">
  <div class=\\"formkit-wrapper\\" data-tab-style=\\"progress\\" data-hide-labels=\\"true\\">
    <div class=\\"formkit-tabs\\" role=\\"tablist\\"><button class=\\"formkit-tab\\" type=\\"button\\" data-active=\\"false\\" data-valid=\\"false\\" data-visited=\\"true\\" role=\\"tab\\" id=\\"input_30_tab_0\\" aria-selected=\\"false\\" aria-controls=\\"input_31\\" tabindex=\\"-1\\" style=\\"\\">
        <!----><span class=\\"formkit-badge\\" role=\\"presentation\\">1</span>
        <!---->
      </button><button class=\\"formkit-tab\\" type=\\"button\\" data-active=\\"true\\" data-valid=\\"true\\" role=\\"tab\\" id=\\"input_30_tab_1\\" aria-selected=\\"true\\" aria-controls=\\"input_33\\" tabindex=\\"0\\" style=\\"\\" data-visited=\\"true\\">
        <!---->
        <!----><span class=\\"formkit-badge\\" role=\\"presentation\\"><!----></span>
      </button><button class=\\"formkit-tab\\" type=\\"button\\" data-active=\\"false\\" data-valid=\\"true\\" role=\\"tab\\" id=\\"input_30_tab_2\\" aria-selected=\\"false\\" aria-controls=\\"input_34\\" tabindex=\\"-1\\" style=\\"\\">
        <!---->
        <!---->
        <!---->
      </button></div>
    <div class=\\"formkit-steps\\">
      <div class=\\"formkit-step\\" data-type=\\"step\\" id=\\"input_31\\" role=\\"tabpanel\\" aria-labelledby=\\"input_30_tab_0\\" hidden=\\"\\">
        <div class=\\"formkit-step-inner\\">
          <div class=\\"formkit-outer\\" data-family=\\"text\\" data-type=\\"text\\" data-invalid=\\"true\\" data-submitted=\\"true\\">
            <div class=\\"formkit-wrapper\\">
              <!---->
              <div class=\\"formkit-inner\\">
                <!---->
                <!----><input class=\\"formkit-input\\" type=\\"text\\" name=\\"text_20\\" id=\\"input_32\\" aria-describedby=\\"input_32-rule_required\\">
                <!---->
                <!---->
              </div>
            </div>
            <!---->
            <ul class=\\"formkit-messages\\">
              <li class=\\"formkit-message\\" id=\\"input_32-rule_required\\" data-message-type=\\"validation\\">Text_20 is required.</li>
            </ul>
          </div>
        </div>
        <div class=\\"formkit-step-actions\\">
          <!---->
          <div class=\\"formkit-step-next\\">
            <div class=\\"formkit-outer\\" data-family=\\"button\\" data-type=\\"button\\">
              <!---->
              <div class=\\"formkit-wrapper\\"><button data-next=\\"true\\" class=\\"formkit-input\\" type=\\"button\\" name=\\"button_21\\" id=\\"input_35\\">
                  <!---->
                  <!---->Next
                  <!---->
                  <!---->
                </button></div>
              <!---->
            </div>
          </div>
        </div>
      </div>
      <div class=\\"formkit-step\\" data-type=\\"step\\" id=\\"input_33\\" role=\\"tabpanel\\" aria-labelledby=\\"input_30_tab_1\\">
        <div class=\\"formkit-step-inner\\">
          <!---->
        </div>
        <div class=\\"formkit-step-actions\\">
          <div class=\\"formkit-step-previous\\">
            <div class=\\"formkit-outer\\" data-family=\\"button\\" data-type=\\"button\\">
              <!---->
              <div class=\\"formkit-wrapper\\"><button data-prev=\\"true\\" class=\\"formkit-input\\" type=\\"button\\" name=\\"button_23\\" id=\\"input_37\\">
                  <!---->
                  <!---->Previous
                  <!---->
                  <!---->
                </button></div>
              <!---->
            </div>
          </div>
          <div class=\\"formkit-step-next\\">
            <div class=\\"formkit-outer\\" data-family=\\"button\\" data-type=\\"button\\">
              <!---->
              <div class=\\"formkit-wrapper\\"><button data-next=\\"true\\" class=\\"formkit-input\\" type=\\"button\\" name=\\"button_22\\" id=\\"input_36\\">
                  <!---->
                  <!---->Next
                  <!---->
                  <!---->
                </button></div>
              <!---->
            </div>
          </div>
        </div>
      </div>
      <div class=\\"formkit-step\\" data-type=\\"step\\" id=\\"input_34\\" role=\\"tabpanel\\" aria-labelledby=\\"input_30_tab_2\\" hidden=\\"\\">
        <div class=\\"formkit-step-inner\\">
          <!---->
        </div>
        <div class=\\"formkit-step-actions\\">
          <div class=\\"formkit-step-previous\\">
            <div class=\\"formkit-outer\\" data-family=\\"button\\" data-type=\\"button\\">
              <!---->
              <div class=\\"formkit-wrapper\\"><button data-prev=\\"true\\" class=\\"formkit-input\\" type=\\"button\\" name=\\"button_24\\" id=\\"input_38\\">
                  <!---->
                  <!---->Previous
                  <!---->
                  <!---->
                </button></div>
              <!---->
            </div>
          </div>
          <!---->
        </div>
      </div>
    </div>
  </div>
</div>"
`;

exports[`multistep Does not allow step advancment when current step is invalid 1`] = `
"<div class=\\"formkit-outer\\" id=\\"input_21\\" data-family=\\"multi-step\\" data-type=\\"multi-step\\">
  <div class=\\"formkit-wrapper\\" data-tab-style=\\"tab\\" data-hide-labels=\\"false\\">
    <div class=\\"formkit-tabs\\" role=\\"tablist\\"><button class=\\"formkit-tab\\" type=\\"button\\" data-active=\\"true\\" data-valid=\\"false\\" data-visited=\\"true\\" role=\\"tab\\" id=\\"input_21_tab_0\\" aria-selected=\\"true\\" aria-controls=\\"input_22\\" tabindex=\\"0\\" style=\\"\\"><span class=\\"formkit-tab-label\\">Step One</span><span class=\\"formkit-badge\\" role=\\"presentation\\">1</span>
        <!---->
      </button><button class=\\"formkit-tab\\" type=\\"button\\" data-active=\\"false\\" data-valid=\\"true\\" role=\\"tab\\" id=\\"input_21_tab_1\\" aria-selected=\\"false\\" aria-controls=\\"input_24\\" tabindex=\\"-1\\" style=\\"\\"><span class=\\"formkit-tab-label\\">Step Two</span>
        <!---->
        <!---->
      </button><button class=\\"formkit-tab\\" type=\\"button\\" data-active=\\"false\\" data-valid=\\"true\\" role=\\"tab\\" id=\\"input_21_tab_2\\" aria-selected=\\"false\\" aria-controls=\\"input_25\\" tabindex=\\"-1\\" style=\\"\\"><span class=\\"formkit-tab-label\\">Step Three</span>
        <!---->
        <!---->
      </button></div>
    <div class=\\"formkit-steps\\">
      <div class=\\"formkit-step\\" data-type=\\"step\\" id=\\"input_22\\" role=\\"tabpanel\\" aria-labelledby=\\"input_21_tab_0\\">
        <div class=\\"formkit-step-inner\\">
          <div class=\\"formkit-outer\\" data-family=\\"text\\" data-type=\\"text\\" data-invalid=\\"true\\" data-submitted=\\"true\\">
            <div class=\\"formkit-wrapper\\">
              <!---->
              <div class=\\"formkit-inner\\">
                <!---->
                <!----><input class=\\"formkit-input\\" type=\\"text\\" name=\\"text_14\\" id=\\"input_23\\" aria-describedby=\\"input_23-rule_required\\">
                <!---->
                <!---->
              </div>
            </div>
            <!---->
            <ul class=\\"formkit-messages\\">
              <li class=\\"formkit-message\\" id=\\"input_23-rule_required\\" data-message-type=\\"validation\\">Text_14 is required.</li>
            </ul>
          </div>
        </div>
        <div class=\\"formkit-step-actions\\">
          <!---->
          <div class=\\"formkit-step-next\\">
            <div class=\\"formkit-outer\\" data-family=\\"button\\" data-type=\\"button\\">
              <!---->
              <div class=\\"formkit-wrapper\\"><button data-next=\\"true\\" class=\\"formkit-input\\" type=\\"button\\" name=\\"button_15\\" id=\\"input_26\\">
                  <!---->
                  <!---->Next
                  <!---->
                  <!---->
                </button></div>
              <!---->
            </div>
          </div>
        </div>
      </div>
      <div class=\\"formkit-step\\" data-type=\\"step\\" id=\\"input_24\\" role=\\"tabpanel\\" aria-labelledby=\\"input_21_tab_1\\" hidden=\\"\\">
        <div class=\\"formkit-step-inner\\">
          <!---->
        </div>
        <div class=\\"formkit-step-actions\\">
          <div class=\\"formkit-step-previous\\">
            <div class=\\"formkit-outer\\" data-family=\\"button\\" data-type=\\"button\\">
              <!---->
              <div class=\\"formkit-wrapper\\"><button data-prev=\\"true\\" class=\\"formkit-input\\" type=\\"button\\" name=\\"button_17\\" id=\\"input_28\\">
                  <!---->
                  <!---->Previous
                  <!---->
                  <!---->
                </button></div>
              <!---->
            </div>
          </div>
          <div class=\\"formkit-step-next\\">
            <div class=\\"formkit-outer\\" data-family=\\"button\\" data-type=\\"button\\">
              <!---->
              <div class=\\"formkit-wrapper\\"><button data-next=\\"true\\" class=\\"formkit-input\\" type=\\"button\\" name=\\"button_16\\" id=\\"input_27\\">
                  <!---->
                  <!---->Next
                  <!---->
                  <!---->
                </button></div>
              <!---->
            </div>
          </div>
        </div>
      </div>
      <div class=\\"formkit-step\\" data-type=\\"step\\" id=\\"input_25\\" role=\\"tabpanel\\" aria-labelledby=\\"input_21_tab_2\\" hidden=\\"\\">
        <div class=\\"formkit-step-inner\\">
          <!---->
        </div>
        <div class=\\"formkit-step-actions\\">
          <div class=\\"formkit-step-previous\\">
            <div class=\\"formkit-outer\\" data-family=\\"button\\" data-type=\\"button\\">
              <!---->
              <div class=\\"formkit-wrapper\\"><button data-prev=\\"true\\" class=\\"formkit-input\\" type=\\"button\\" name=\\"button_18\\" id=\\"input_29\\">
                  <!---->
                  <!---->Previous
                  <!---->
                  <!---->
                </button></div>
              <!---->
            </div>
          </div>
          <!---->
        </div>
      </div>
    </div>
  </div>
</div>"
`;
